worlds/ffmq/__init__.py

```diff
@@ -0,0 +1,209 @@
+        ["Alchav"]
+    )]
+
+
+class FFMQWorld(World):
+    """Final Fantasy: Mystic Quest is a simple, humorous RPG for the Super Nintendo. You travel across four continents,
+    linked in the middle of the world by the Focus Tower, which has been locked by four magical coins. Make your way to
+    the bottom of the Focus Tower, then straight up through the top!"""
+    # -Giga Otomia
+
+    game = "Final Fantasy Mystic Quest"
+
+    item_name_to_id = {name: data.id for name, data in item_table.items() if data.id is not None}
+    location_name_to_id = location_table
+    option_definitions = option_definitions
+
+    topology_present = True
+
+    item_name_groups = item_groups
+
+    generate_output = generate_output
+    create_items = create_items
+    create_regions = create_regions
+    set_rules = set_rules
+
+    data_version = 1
+    
+    web = FFMQWebWorld()
+    # settings: FFMQSettings
+
+    def __init__(self, world, player: int):
+        self.rom_name_available_event = threading.Event()
+        self.rom_name = None
+        self.rooms = None
+        super().__init__(world, player)
+
+    def generate_early(self):
+        if self.multiworld.sky_coin_mode[self.player] == "shattered_sky_coin":
+            self.multiworld.brown_boxes[self.player] = self.multiworld.brown_boxes[self.player].from_text("include")
+        if self.multiworld.enemies_scaling_lower[self.player].value > \
+                self.multiworld.enemies_scaling_upper[self.player].value:
+            (self.multiworld.enemies_scaling_lower[self.player].value,
+             self.multiworld.enemies_scaling_upper[self.player].value) =\
+                (self.multiworld.enemies_scaling_upper[self.player].value,
+                 self.multiworld.enemies_scaling_lower[self.player].value)
+        if self.multiworld.bosses_scaling_lower[self.player].value > \
+                self.multiworld.bosses_scaling_upper[self.player].value:
+            (self.multiworld.bosses_scaling_lower[self.player].value,
+             self.multiworld.bosses_scaling_upper[self.player].value) =\
+                (self.multiworld.bosses_scaling_upper[self.player].value,
+                 self.multiworld.bosses_scaling_lower[self.player].value)
+
+    @classmethod
+    def stage_generate_early(cls, multiworld):
+
+        # api_urls = Utils.get_options()["ffmq_options"].get("api_urls", None)
+        api_urls = [
+            "https://api.ffmqrando.net/",
+            "http://ffmqr.jalchavware.com:5271/"
+        ]
+
+        rooms_data = {}
+
+        for world in multiworld.get_game_worlds("Final Fantasy Mystic Quest"):
+            if (world.multiworld.map_shuffle[world.player] or world.multiworld.crest_shuffle[world.player] or
+                    world.multiworld.crest_shuffle[world.player]):
+                if world.multiworld.map_shuffle_seed[world.player].value.isdigit():
+                    multiworld.random.seed(int(world.multiworld.map_shuffle_seed[world.player].value))
```

Comment:
Using the `MultiWorld.random` in stages is nondeterministic. Should probably generate the seeds here from the random object of one of the players.

Reply:
This is not factually correct. It only isn't deterministic during generate_output.

Reply:
then we would have bigger problems, no? https://github.com/ArchipelagoMW/Archipelago/blob/d49860fbebcbde28774ae091a1505290d430b505/worlds/alttp/__init__.py#L627

Reply:
`stage_callable` is non-deterministic, but it apparently only gets called for like 3 methods so berserker is correct actually. Core has a lot of non-recycled code which can be misleading.